k2/csrc/cuda/log_test.cu

```diff
@@ -0,0 +1,114 @@
+// k2/csrc/cuda/log.h
+
+// Copyright (c)  2020  Fangjun Kuang (csukuangfj@gmail.com)
+
+// See ../../../LICENSE for clarification regarding multiple authors
+
+#include "gtest/gtest.h"
+#include "k2/csrc/cuda/log.h"
+
+namespace k2 {
+
+TEST(Log, Cpu) {
+  K2_LOG(DEBUG) << "Debug message";
+  K2_LOG(INFO) << "Info message";
+  K2_LOG(WARNING) << "Warning message";
+  K2_LOG(ERROR) << "Error message";
+
+  K2_DLOG(INFO) << "This is printed only in debug mode";
+
+  int32_t a = 10;
+  int32_t b = 20;
+  int32_t c = 30;
+  int32_t d = a;
+
+  K2_DCHECK_EQ(a, d) << "This is checked only in debug mode";
+
+  K2_CHECK(a == d) << "failed";
+
+  K2_CHECK_EQ(a, d) << "failed";
+  K2_CHECK_NE(a, b) << "failed";
+
+  K2_CHECK_LE(a, b) << "failed";
+  K2_CHECK_LT(a, b) << "failed";
+
+  K2_CHECK_GE(c, b) << "failed";
+  K2_CHECK_GT(c, b) << "failed";
+}
+
+__global__ void DummyKernel(int32_t *b, int32_t a) {
+  K2_DLOG(INFO) << "In kernel";
+  K2_DCHECK_LT(*b, a);  // enabled only in debug mode
+  *b += 1;
+  K2_CHECK_EQ(*b, a);
+  K2_DLOG(DEBUG) << "Done";
+}
+
+TEST(Log, Cuda) {
+  K2_LOG(INFO) << "Test log for cuda";
+  int32_t a = 10;
+  int32_t *b;
+  auto ret = cudaMalloc(&b, sizeof(a));
+  K2_CHECK_EQ(ret, cudaSuccess) << "Failed to allocate memory";
+
+  ret = cudaMemcpy(b, &a, sizeof(a), cudaMemcpyHostToDevice);
+  K2_CHECK_EQ(ret, cudaSuccess) << "Failed to copy memory to gpu";
+
+  DummyKernel<<<1, 1>>>(b, a + 1);
+
+  int32_t c = 0;
+  ret = cudaMemcpy(&c, b, sizeof(a), cudaMemcpyDeviceToHost);
+  K2_CHECK_CUDA_ERROR(ret) << "Failed to copy memory from gpu";
+  K2_CHECK_EQ(a + 1, c) << "Error in the kernel!";
+
+  ret = cudaFree(b);
+  K2_DCHECK_CUDA_ERROR(ret) << "Failed to free gpu memory";
+}
+
+TEST(LogDeathTest, NegativeCases) {
+  ASSERT_DEATH(K2_LOG(FATAL) << "This will crash the program", "");
+
+  int32_t a = 10;
+  int32_t b = 20;
+  int32_t c = a;
+  ASSERT_DEATH(K2_CHECK_EQ(a, b), "");
+  ASSERT_DEATH(K2_CHECK_NE(a, c), "");
+
+  ASSERT_DEATH(K2_CHECK_LE(b, a), "");
+  ASSERT_DEATH(K2_CHECK_LT(b, a), "");
+
+  ASSERT_DEATH(K2_CHECK_GE(a, b), "");
+  ASSERT_DEATH(K2_CHECK_GT(a, b), "");
+
+  auto ret = cudaErrorMemoryAllocation;
+  ASSERT_DEATH(K2_CHECK_CUDA_ERROR(ret), "");
+
+  ret = cudaErrorAssert;
+  ASSERT_DEATH(K2_CHECK_CUDA_ERROR(ret), "");
+
+  // NOTE: normally we do not need to
+  // check if NDEBUG is defined in order
+  // to use K2_DCHECK_*. ASSERT_DEATH
+  // expects that the statement will make
+  // the program crash and this is only
+  // possible for the debug build,
```

Comment:
a little confused here? `ASSERT_DEATH` works just for debug build? I noticed that you call `ASSERT_DEATH` above without debug guard

Reply:
For the release build, K2_DCHECK_* expands to empty, but `ASSERT_DEATCH` expects that `K2_DCHECK_*` will
kill the thread but actually it does not in the release build, so the test will fail.

Reply:
If you remove the guard, the unit test will fail in the debug build.

Reply:
Oh, got it, thanks